e2e/framework/framework.go

```diff
@@ -127,23 +215,31 @@ func (f *Framework) runSuite(t *testing.T, s *TestSuite) (skip bool, err error)
 		}
 	}
 
+	info, err := f.provisioner.SetupTestSuite(t, provisioning.SetupOptions{
+		Name:         s.Component,
+		ExpectConsul: s.Consul,
+		ExpectVault:  s.Vault,
+	})
+	if err != nil {
+		t.Fatalf("could not provision cluster: %v", err)
+	}
+	defer f.provisioner.TearDownTestSuite(t, info.ID)
+
 	for _, c := range s.Cases {
 		// The test name is set to the name of the implementing type, including package
 		name := fmt.Sprintf("%T", c)
 
-		// Each TestCase is provisioned a Nomad cluster to test against.
-		// This varies by the provisioner implementation used, currently
-		// the default behavior is for every Test case to use a single shared
-		// Nomad cluster.
-		info, err := f.provisioner.ProvisionCluster(ProvisionerOptions{
+		// The ClusterInfo handle should be used by each TestCase to isolate
+		// job/task state created during the test.
+		info, err := f.provisioner.SetupTestCase(t, provisioning.SetupOptions{
 			Name:         name,
 			ExpectConsul: s.Consul,
 			ExpectVault:  s.Vault,
 		})
 		if err != nil {
 			return false, fmt.Errorf("could not provision cluster for case: %v", err)
 		}
-		defer f.provisioner.DestroyCluster(info.ID)
+		defer f.provisioner.TearDownTestCase(t, info.ID)
```

Comment:
`defer` inside a loop is often a bug (and linters warn against it) - previously `DestroyCluster` was a noop and so it didn't really matter, but now I'm thinking we actually want to tear down each test case after it runs, otherwise I think we're putting the provisioned cluster into a kind of accumulated state as each Case runs

Reply:
Yup, you're totally right. Will fix.

Reply:
Fixed in aa3b15d. I want to make sure we're still `defer`ing the teardown so that `t.Fatal` doesn't prevent the teardown step from happening, so I factored this inner loop out into a `runCase` function so that we run the deferred teardown at the end of that scope.